### ISSUE TITLE
Cannot log in when selecting the GitLab back-end (sporadically) 

### ISSUE THREAD

**Issue by munen** [200ok-ch/organice#739]
**Describe the bug**

@branch14, the co-founder of 200ok.ch, was super happy to hear about the GitLab back-end in organice, so he wanted to use it right away. At first, he could login using Firefox on the Desktop and could log in on his smartphone (Android).

After some time passed, he got an error on organice on the smartphone, so he tried logging back in. Now, logging in on the smartphone reproducibly doesn't work whereas it did work before - and he's still logged in on the Desktop browser.

**To Reproduce**
0. Use @branch14 Android device (yes, this is where the reproducability already stops unfortunately)
1. Go to https://organice.200ok.ch
2. Click on 'Sign in'
3. Select 'GitLab', enter his org repo

**Screenshots**

![Screenshot_20211102_185224_org mozilla fenix](https://user-images.githubusercontent.com/505721/139920693-99952f1f-f2e3-4321-b964-c23c5fd57fd9.jpg)

The visible error is generated here: https://github.com/200ok-ch/organice/blob/master/src/actions/sync_backend.js#L69

When inspecting the error by connecting the phone to his Desktop, we see that the issue is not actually 'retrieving the directory', but one step ealier while authenticating against GitLab.

![image](https://user-images.githubusercontent.com/505721/139919357-a1c52655-61fd-44d5-bf03-6b4b6f38eddb.png)

![image](https://user-images.githubusercontent.com/505721/139919390-741030a9-7c85-4de9-a32d-b1620ebe322a.png)

**Smartphone (please complete the following information):**
 - Device: Huawei P20 Pro
 - OS: I'm assuming some current Android, don't think/hope that's relevant.
 - Browser: Firefox

**Comment by munen** [comment]
@chasecaleb Do you have an idea what this is about?

The only additional context that I can provide that @branch14 can can continue to use organice.200ok.ch on the Desktop and while he is describing the issue, I can continue to log out and in both on the Desktop and my iPhone, so it's likely not a general outage on the side of GitLab.

**Comment by chasecaleb** [comment]
Ugh, sorry about this! So just to make sure I'm on the same page, there's two scenarios where that token call can happen:
1. While obtaining the initial token, i.e. clicking "sign in" from https://organice.200ok.ch/sign_in
2. While using the refresh token to obtain a new access token, i.e. at some random time while using the app (roughly >= 2 hours after sign in)

It sounds like there's an occur occurring for both scenarios, right? Since @branch14 was using the app happy for a while, then scenario 2 caused an occur, then he logged out and attempted to sign in again but encountered scenario 1.

---

I'm going to do some more digging to figure out why this is going on and I'll get back to you within 12-36 hours, but in the meantime @munen can you post a screenshot of how the OAuth application is configured in GitLab, please? Go to https://gitlab.com/-/profile/applications , then click the edit button (pencil icon) in the "Your applications" section towards the end of the page. For example this is how I have my test application configured:

![](file:///tmp/2021-11-02-133514_1029x1342_scrot.png)

**Comment by munen** [comment]
> Ugh, sorry about this!

Not a problem at all. And even if it were, you've responded within 30 minutes from across the world - that's the most expensive SLA I can think of :rocket: :superhero: 

> It sounds like there's an occur occurring for both scenarios, right?

This is very much how I understood it while he explained it to me. But @branch14 is the person who can confirm definitively. I can confirm that variant 1 (sign in) is occuring now as I've seen him debug it during a screen sharing session. Since he cannot log in anymore (from the smartphone), it'll be hard to debug variant 2. But I'll make sure to ping him off-channel so he can confirm about the issue in variant 2.

> I'm going to do some more digging to figure out why this is going on and I'll get back to you within 12-36 hours

Amazing, thanks :bow: :pray: 

> @munen can you post a screenshot of how the OAuth application is configured in GitLab, please

Sure, I'm happy to help!

![image](https://user-images.githubusercontent.com/505721/139925566-fda38a54-b345-451c-853c-94ccec12b4be.png)

**Comment by chasecaleb** [comment]
@branch14 if you're able to run this for me via remote dev tools and share the output I would appreciate it:

```javascript
      (() => {
          const state = JSON.parse(localStorage["oauth2authcodepkce-state"]);
          console.log(Object.keys(state));
          console.log(Object.keys(state.accessToken));
          console.log(state.accessToken?.expiry || 'missing');
          console.log(state.hasAuthCodeBeenExchangedForAccessToken);
          console.log(Object.keys(state.refreshToken));
          console.log(localStorage.authenticatedSyncService);
          console.log(window.location.origin);
      })();
```

Text or screenshot is fine, whichever is more convenient for you.

**Comment by branch14** [comment]
@chasecaleb Thx for looking into this!

![organice_gitlab](https://user-images.githubusercontent.com/20964/139940449-6243bcd7-0051-44bc-b202-254a584f5e5c.png)

**Comment by branch14** [comment]
I can confirm that @munen described the issue I have.

Case 2 might have been the initial occurrence of the issue, but I can only reproduce case 1, as I cannot successfully login, atm.

**Comment by chasecaleb** [comment]
Update: I talked @branch14 on IRC and we found a workaround:
1. Clear `localStorage`. He did this via `localStorage.clear()` from a remote dev tools console, but I'm sure there's easier ways to do it on-device too.
2. Sign in again

I think the issue occurred because `localStorage.authenticatedSyncService` was cleared somehow, but so far I'm not sure what bit of code caused that to happen. I'm still looking into the root cause.
